sweep/fee_bumper.go

```diff
@@ -76,4 +92,436 @@ type BumpResult struct {
+func (t *TxPublisher) calcCurrentConfTarget(deadline int32) uint32 {
+	var confTarget uint32
+
+	// Calculate how many blocks left until the deadline.
+	deadlineDelta := deadline - t.currentHeight
+
+	// If we are already past the deadline, we will set the conf target to
+	// be 0.
+	if deadlineDelta < 0 {
+		log.Warnf("Deadline is %d blocks behind current height %v",
+			-deadlineDelta, t.currentHeight)
+
+		confTarget = 0
+	} else {
+		confTarget = uint32(deadlineDelta)
+	}
+
+	return confTarget
+}
+
+// calcSweepTxWeight calculates the weight of the sweep tx. It assumes a
+// sweeping tx always has a single output(change).
+func (t *TxPublisher) calcSweepTxWeight(inputs []input.Input,
+	output []byte) (uint64, error) {
+
+	// Use a const fee rate as we only use the weight estimator to
+	// calculate the size.
+	const feeRate = 1
+
+	// Initialize the tx weight estimator with,
+	// - nil outputs as we only have one single change output.
+	// - const fee rate as we don't care about the fees here.
+	// - 0 maxfeerate as we don't care about fees here.
+	//
+	// TODO(yy): we should refactor the weight estimator to not require a
+	// fee rate and max fee rate and make it a pure tx weight calculator.
+	_, estimator, err := getWeightEstimate(
+		inputs, nil, feeRate, 0, output,
+	)
+	if err != nil {
+		return 0, err
+	}
+
+	return uint64(estimator.weight()), nil
+}
+
+// createRBFCompliantTx creates a tx that is compliant with RBF rules. It does
+// so by creating a tx, validate it using `TestMempoolAccept`, and bump its fee
+// and redo the process until the tx is valid, or return an error when non-RBF
+// related errors occur or the budget has been used up.
+func (t *TxPublisher) createRBFCompliantTx(req *BumpRequest,
+	f FeeFunction) (uint64, error) {
+
+	for {
+		// Create a new tx with the given fee rate and check its
+		// mempool acceptance.
+		tx, fee, err := t.createAndCheckTx(
+			req.Inputs, req.ChangeScript, f,
+		)
+
+		switch {
+		case err == nil:
+			// Increase the request counter.
+			//
+			// NOTE: this is the only place where we increase the
+			// counter.
+			requestID := t.requestCounter.Add(1)
+
+			// Register the record.
+			t.records.Store(requestID, &monitorRecord{
+				tx:           tx,
+				set:          req.Inputs,
+				changeScript: req.ChangeScript,
+				feeFunction:  f,
+				fee:          fee,
+			})
+
+			return requestID, nil
+
+		// If the error indicates the fees paid is not enough, we will
+		// ask the fee function to increase the fee rate and retry.
+		//
+		// TODO(yy): add this new error in btcwallet to map RBF related
+		// reject reasons:
+		// case errors.Is(err, lnwallet.ErrInsufficientFee):
+		case errors.Is(err, lnwallet.ErrMempoolFee):
+			// If the fee function tells us that we have used up
+			// the budget, we will return an error indicating this
+			// tx cannot be made. The sweeper should handle this
+			// error and try to cluster these inputs differetly.
+			err = f.IncreaseFeeRate()
```

Comment:
Ok now I see how this is working. Feels like rather than converging on this iteratively, can't we just find out what the mempool purge rate is and choose the greater of what comes out of the fee function and the mempool floor?

Reply:
> can't we just find out what the mempool purge rate is and choose the greater of what comes out of the fee function and the mempool floor?

Due to higher order effects of the fee filter, we can't reliably ascertain what the "global purge" rate is. If all our peers have high fee filters, then we won't see a large class of txns, causing us to potentially never have to purge at all.

Reply:
Fair enough, but how does this environmental mutation fix that? We still only have a local view of the mempool anyway and this bumping directive is buried inside of a recognition of ErrMempoolFee, which idk how we ascertain.

Reply:
We need to match more errors here - initially I tried with `ErrReplacement`, then realized we only want to match the case when it's indeed the violation of fee-related rules, not others, so I created more granular error matching [here](https://github.com/yyforyongyu/btcwallet/blob/unify-error-handling/rpcclient/errors.go#L217), need to open a PR on the `btcwallet` side.